src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_Warnings.cs

```diff
@@ -719,6 +720,45 @@ private void CheckLiftedUnaryOp(BoundUnaryOperator node)
             }
         }
 
+        private void CheckBinOp(BoundBinaryOperator node)
+        {
+            Debug.Assert(node != null);
+
+            if (node.OperatorKind.IsLifted())
+            {
+                return;
+            }
+
+            if (!_compilation.FeatureStrictEnabled)
+            {
+                return;
+            }
+
+            switch (node.OperatorKind.Operator())
+            {
+                case BinaryOperatorKind.Equal:
+                case BinaryOperatorKind.NotEqual:
+                    // CS0472: The result of the expression is always '{0}' since a value of type '{1}' is never equal to 'null' of type '{2}'
+                    //
+                    // Produce the warning if one side is always null and the other is never null.
+                    // That is, we have something like "if (myInt == null)"
+
+                    string always = node.OperatorKind.Operator() == BinaryOperatorKind.NotEqual ? "true" : "false";
+                    TypeSymbol valueType;
+
+                    // we use a separate warning code for cases newly detected in later versions of the compiler
+                    if (node.Right.IsLiteralOrConvertedNull() && node.Left.IsBoxedValueType(out valueType))
+                    {
+                        Error(ErrorCode.WRN_NubExprIsConstBool2, node, always, valueType, node.Right.Type);
```

Comment:
nit: we could probably factor the `IsBoxedValueType` check and the `Error` call by checking which side is "null" and storing the other side in a variable. Then we can proceed unconditionally.

Reply:
Yes, I agree, we should refactor the code to avoid redundancy, but there are several similar methods here that all use the same pattern. We should refactor all or none.

---
In reply to: [205677618](https://github.com/dotnet/roslyn/pull/28869#discussion_r205677618) [](ancestors = 205677618)